process_report/process_report.py

```diff
@@ -112,21 +172,99 @@ def remove_billables(dataframe, pi, projects, output_file):
 
     So this *keeps* the projects/pis that should not be billed.
     """
-    filtered_dataframe = dataframe[dataframe['Manager (PI)'].isin(pi) | dataframe['Project - Allocation'].isin(projects)]
+    filtered_dataframe = dataframe[dataframe[PI_FIELD].isin(pi) | dataframe[PROJECT_FIELD].isin(projects)]
     filtered_dataframe.to_csv(output_file, index=False)
 
+
 def export_pi_billables(dataframe: pandas.DataFrame, output_folder):
     if not os.path.exists(output_folder):
         os.mkdir(output_folder)
 
-    invoice_month = dataframe['Invoice Month'].iat[0]
-    pi_list = dataframe['Manager (PI)'].unique()
+    invoice_month = dataframe[INVOICE_DATE_FIELD].iat[0]
+    pi_list = dataframe[PI_FIELD].unique()
 
     for pi in pi_list:
-        pi_projects = dataframe[dataframe['Manager (PI)'] == pi]
-        pi_instituition = pi_projects['Institution'].iat[0]
+        if pandas.isna(pi): 
+            continue
+        pi_projects = dataframe[dataframe[PI_FIELD] == pi]
+        pi_instituition = pi_projects[INSTITUTION_FIELD].iat[0]
         pi_projects.to_csv(output_folder + f"/{pi_instituition}_{pi}_{invoice_month}.csv")
-        
+
+
+def apply_credits_new_pi(dataframe, old_pi_file):
+    """Applies the New PI Credit. This credit function expects the 
+    command arguement `old-pi-file` to be set, pointing to a txt file containing old PIs"""
+    new_pi_credit_code = "0002"
+    new_pi_credit_amount = 1000
+
+    old_pi_list = set()
+
+    try:
+        with open(old_pi_file) as f:
+            for pi in f: 
+                old_pi_list.add(pi.strip())
+
+    except FileNotFoundError:
+        print("Applying credit 0002 failed. Old PI file does not exist")
+        sys.exit(1)
+
+    pi_list = dataframe[PI_FIELD].unique()
+
+    for pi in pi_list:
+        if pandas.isna(pi): 
+            continue   # NaN check
+        if pi in old_pi_list: 
```

Comment:
This function does too much. I would split this out into its own function `is_new_pi` that returns a bool whether the PI is old or new. Move over the logic for the file reading, etc into that too.

Reply:
You want the file reading and the pi checking to be bundled in one function? I'm having a hard time to see how to do that.

Reply:
Two functions then. You can have a function `load_old_pis(old_pi_file) -> dict()` and `is_new_pi(str: username, str: invoice_month) -> bool`

Reply:
Additionally, if we want to be able to run this program reliably, the PI file will need to contain the first invoice month of a PI, otherwise the execution of the program will not be idempotent in the case where this program will update the PI file as it executes.

Reply:
Ok, I'll make it so that PIs whose first invoice month is the same as the month of the invoice being processed will be counted as new PIs for the purpose of applying the credit.